server/src/main/java/org/apache/cassandra/sidecar/acl/authentication/ReloadingJwtAuthenticationHandler.java

```diff
@@ -105,10 +108,18 @@ public void authenticate(RoutingContext context, Handler<AsyncResult<User>> hand
             }
 
             List<String> roles = extractCassandraRoles(decodedToken);
-            if (!roles.isEmpty())
+            String roleIntended = context.request().getHeader(AUTH_ROLE);
+
+            if (isNotEmpty(roleIntended) && !roles.contains(roleIntended))
             {
-                user.attributes().put(CASSANDRA_ROLES_ATTRIBUTE_NAME, roles);
+                String errMsg = String.format("User not authorized for role %s", roleIntended);
+                handler.handle(Future.failedFuture(wrapHttpException(UNAUTHORIZED, errMsg)));
+                return;
             }
+
+            List<String> rolesToAdd = roleIntended != null && !roleIntended.isEmpty()
```

Comment:
NIT: since you are using isNotEmpty in this class already
```suggestion
            List<String> rolesToAdd = isNotEmpty(roleIntended)
```